blinky/ui/src/main/java/no/nordicsemi/android/blinky/control/viewmodel/BlinkyViewModel.kt

```diff
@@ -0,0 +1,92 @@
+package no.nordicsemi.android.blinky.control.viewmodel
+
+import android.app.Application
+import android.content.Context
+import android.net.Uri
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.viewModelScope
+import dagger.hilt.android.lifecycle.HiltViewModel
+import dagger.hilt.android.qualifiers.ApplicationContext
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.launch
+import no.nordicsemi.android.blinky.control.BlinkyDestination
+import no.nordicsemi.android.blinky.control.BlinkyParams
+import no.nordicsemi.android.blinky.control.R
+import no.nordicsemi.android.blinky.control.repository.BlinkyRepository
+import no.nordicsemi.android.common.logger.NordicLogger
+import no.nordicsemi.android.common.navigation.NavigationManager
+import no.nordicsemi.android.log.LogContract.Log
+import no.nordicsemi.android.log.timber.nRFLoggerTree
+import timber.log.Timber
+import javax.inject.Inject
+
+@HiltViewModel
+class BlinkyViewModel @Inject constructor(
+    @ApplicationContext context: Context,
+    private val repository: BlinkyRepository,
+    navigationManager: NavigationManager,
+) : AndroidViewModel(context as Application) {
+    /** The Blinky device name, as advertised. */
+    val deviceName: String
+    /** The connection state of the device. */
+    val state = repository.state
+    /** The LED state. */
+    val ledState = repository.ledState
+        .map { it.also { Timber.log(Log.Level.APPLICATION, "LED state changed to $it") } }
+        .stateIn(viewModelScope, SharingStarted.Lazily, false)
+    /** The button state. */
+    val buttonState = repository.buttonState
+        .map { it.also { Timber.log(Log.Level.APPLICATION, "Button state changed to $it") } }
+        .stateIn(viewModelScope, SharingStarted.Lazily, false)
+
+    /** Timber tree that logs to nRF Logger. */
+    private val tree: Timber.Tree
+    /** If the nRF Logger is installed, this URI will allow to open the session. */
+    private val sessionUri: Uri?
+
+    init {
+        // Get the navigation arguments.
+        val parameters = navigationManager.getArgument(BlinkyDestination) as BlinkyParams
+
+        // Plant a new Tree that logs to nRF Logger.
+        val key = parameters.device.address
+        val name = parameters.deviceName ?: context.getString(R.string.unnamed_device)
+        tree = nRFLoggerTree(context, null, key, name)
+            .also { Timber.plant(it) }
+            .also { sessionUri = it.session?.sessionUri }
+
+        // Update the device name.
+        deviceName = name
+
+        connect()
+    }
+
+    fun connect() {
+        val exceptionHandler = CoroutineExceptionHandler { _, _ -> }
```

Comment:
I still don't understand the purpose of this CoroutineExceptionHandler? Are those suppose to prevent crashes of the other Jobs? I think this is the job of SupervisorJob() and viewModelScope already have this. Maybe you should print an exception here. Otherwise I think it can be removed.

Reply:
It is needed. This, or try-catch around `connect()`.